File: js130/challenges/medium/simple-linked-list.js
class SimpleLinkedList {
  constructor() {
    this.headElement = undefined;
  }

  static fromArray(array) {
    let list = new SimpleLinkedList();
    if (!array || array.length === 0) return list;
    let reversedArray = array.slice().reverse();
    reversedArray.forEach(element => list.push(element));
    return list;
  }

  size() {
    let size = 0;
    let currentElement = this.headElement;
    while (currentElement) {
      currentElement = currentElement.next();
      size += 1;
    }
    return size;
  }

  isEmpty() {
    return this.size() === 0;
  }

  push(storedDatum) {
    let headElement = new Element(storedDatum, this.headElement);
    this.headElement = headElement;
  }

  peek() {
    return this.headElement ? this.headElement.datum() : null;
  }

  head() {
    return this.headElement;
  }

  pop() {
    let previousHead = this.headElement;
    this.headElement = previousHead.next();
    return previousHead.datum();
  }

  toArray() {
    let array = [];
    let currentElement = this.headElement;
    while (currentElement) {
      array.push(currentElement.datum());
      currentElement = currentElement.next();
    }
    return array;
  }

  reverse() {
    let currentElement = this.headElement;
    let nextElement = this.headElement.next();
    currentElement.nextElement = undefined;

    while (nextElement) {
      let nextElementNextElement = nextElement.next();
      nextElement.nextElement = currentElement;
      currentElement = nextElement;
      nextElement = nextElementNextElement;
    }

    this.headElement = currentElement;

    return this;

    // return SimpleLinkedList.fromArray(this.toArray().reverse()); this code creates new objects so it is memory inefficient
  }
}

class Element {
  constructor(storedDatum, nextElement) {
    this.storedDatum = storedDatum;
    this.nextElement = nextElement;
  }

  datum() {
    return this.storedDatum;
  }

  isTail() {
    return !this.next();
  }

  next() {
    return this.nextElement || null;
  }
}

module.exports = { SimpleLinkedList, Element };

File: py130/challenges/medium_5/simple_linked_list.py
class Element:
    def __init__(self, datum, next=None):
        self.datum = datum
        self.next = next

    def is_tail(self):
        return self.next == None

class SimpleLinkedList:
    def __init__(self):
        self.head = None

    @property
    def size(self):
        count = 0

        next_element = self.head

        while next_element:
            count += 1
            next_element = next_element.next

        return count

    def is_empty(self):
        return self.size == 0

    def push(self, datum):
        element = Element(datum, self.head)
        self.head = element

    def peek(self):
        if self.head:
            return self.head.datum
        else:
            return None

    def pop(self):
        popped = self.head
        self.head = self.head.next
        return popped.datum

    @classmethod
    def from_list(cls, lst):

        new_lst = cls()

        if lst:
            for item in reversed(lst):
                new_lst.push(item)

        return new_lst

    def to_list(self):
        lst = []

        next_element = self.head

        while next_element:
            lst.append(next_element.datum)
            next_element = next_element.next

        return lst

    def reverse(self):
        lst = self.to_list()
        lst = list(reversed(lst))
        return self.__class__.from_list(lst)